resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala

```diff
@@ -267,8 +334,79 @@ private[spark] class KubernetesClusterSchedulerBackend(
     }
     true
   }
+
+  override def createDriverEndpoint(properties: Seq[(String, String)]): DriverEndpoint = {
+    new KubernetesDriverEndpoint(rpcEnv, properties)
+  }
+
+  private class KubernetesDriverEndpoint(rpcEnv: RpcEnv, sparkProperties: Seq[(String, String)])
+    extends DriverEndpoint(rpcEnv, sparkProperties) {
+    override def receiveAndReply(context: RpcCallContext): PartialFunction[Any, Unit] = {
+      new PartialFunction[Any, Unit]() {
+        override def isDefinedAt(msg: Any): Boolean = {
+          msg match {
+            case RetrieveSparkAppConfig(executorId) =>
+              Utils.isDynamicAllocationEnabled(sc.conf)
+            case _ => false
+          }
+        }
+
+        override def apply(msg: Any): Unit = {
+          msg match {
+            case RetrieveSparkAppConfig(executorId) =>
+              EXECUTOR_MODIFICATION_LOCK.synchronized {
+                var resolvedProperties = sparkProperties
+                val runningExecutorPod = kubernetesClient
+                  .pods()
+                  .withName(runningExecutorPods(executorId).getMetadata.getName)
+                  .get()
+                val nodeName = runningExecutorPod.getSpec.getNodeName
+                val shuffleServicePods = kubernetesClient
+                  .pods()
+                  .inNamespace(shuffleServiceConfig.get.shuffleNamespace)
+                  .withLabels(shuffleServiceConfig.get.shuffleLabels.asJava)
+                  .list()
+
+                // Filter only those pods which are posting Ready status and on the same
+                // node as our executor.
+                val readyPods = shuffleServicePods
+                  .getItems.asScala.toList.filter(pod => Readiness.isPodReady(pod))
+                val colocatedPods = readyPods.filter(_.getSpec.getNodeName == nodeName)
```

Comment:
what happens if the executor pod has been rescheduled to a different node between when we checked it in line 363 and here?  Maybe executor pod should also check that the shuffle service it's connecting to is indeed colocated with it.

This also brings up the question of executor IDs.  I think we need to hold that when an executor is restarted it gets a new ID, so the idea is that the old one died and the new one was created, rather than the old one was moved.  We don't want to have to move executor state with the rescheduled pod.

With that assumption then that executor IDs once assigned always refer to the same JVM, we can begin working towards the resilient executor logic.